ygot/pathstrings.go

```diff
@@ -0,0 +1,268 @@
+			gp, err := StringToStringSlicePath(path)
+			if err != nil {
+				errs = util.AppendErr(errs, fmt.Errorf("error building string slice path: %v", err))
+				continue
+			}
+			pmsg.Element = gp.Element
+		}
+	}
+
+	if errs != nil {
+		return nil, errs
+	}
+
+	return pmsg, nil
+}
+
+// StringToStringSlicePath takes a string representing a path, and converts it into a
+// gnmi.Path. For example, if the Path "/a/b[c=d]/e" is input, it is converted
+// to a gnmi.Path{Element: []string{"a", "b[c=d]", "e"}} which is returned.
+func StringToStringSlicePath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+	for _, p := range parts {
+		// Run through extractKV to ensure that the path is valid.
+		if _, _, err := extractKV(p); err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+	}
+
+	return &gnmipb.Path{
+		Element: parts,
+	}, nil
+}
+
+// StringToStructuredPath takes a string representing a path, and converts it to
+// a gnmi.Path, using the PathElem element message that is defined in gNMI 0.4.0.
+func StringToStructuredPath(path string) (*gnmipb.Path, error) {
+	parts := pathStringToElements(path)
+
+	gpath := &gnmipb.Path{}
+	for _, p := range parts {
+		name, kv, err := extractKV(p)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing path %s: %v", path, err)
+		}
+		gpath.Elem = append(gpath.Elem, &gnmipb.PathElem{
+			Name: name,
+			Key:  kv,
+		})
+	}
+	return gpath, nil
+}
+
+// pathStringToElements splits the string s, which represents a gNMI string
+// path into its constituent elements. It does not parse keys, which are left
+// unchanged within the path - but removes escape characters from element
+// names. The path returned omits any leading empty elements when splitting
+// on the / character.
+func pathStringToElements(s string) []string {
+	var parts []string
+	var buf bytes.Buffer
+
+	var inKey, inEscape bool
+
+	for _, ch := range s {
+		switch {
+		case ch == '[' && !inEscape:
+			inKey = true
+		case ch == ']' && !inEscape:
+			inKey = false
+		case ch == '\\' && !inEscape && !inKey:
+			inEscape = true
+			continue
+		case ch == '/' && !inEscape && !inKey:
+			parts = append(parts, buf.String())
+			buf.Reset()
+			continue
+		}
+
+		buf.WriteRune(ch)
+		inEscape = false
+	}
+
+	if buf.Len() != 0 {
+		parts = append(parts, buf.String())
+	}
+
+	if len(parts) > 0 && parts[0] == "" {
+		parts = parts[1:]
+	}
+
+	return parts
+}
+
+// extractKV extracts key value predicates from the input string in. It returns
+// the name of the element, a map keyed by key name with values of the predicates
+// specified. It removes escape characters from keys and values where they are
+// specified.
+func extractKV(in string) (string, map[string]string, error) {
+	var inEscape, inKey, inValue bool
+	var name, currentKey string
+	var buf bytes.Buffer
+	keys := map[string]string{}
+
+	for _, ch := range in {
+		switch {
+		case ch == '[' && !inEscape && !inKey:
+			inKey = true
```

Comment:
Since it would cost little and add a layer of verification I suggest some additional simple checks e.g. here, if inKey -> error. I could be reading this wrong but it looks like the current code would allow [[.

Reply:
Added this check and a test.

Reply:
Added a check for unescaped ] in an element name, are there others that you think are missing here? (This is getting very esoteric - since the specification is very loose here.)